invenio_files_rest/views.py

```diff
@@ -50,6 +55,92 @@
 )
 
 
+def pass_bucket(f):
+    """Decorator to retrieve a bucket."""
+    @wraps(f)
+    def decorate(*args, **kwargs):
+        bucket_id = kwargs.pop('bucket_id')
+        bucket = Bucket.get(bucket_id)
+        if not bucket:
+            abort(404, 'Bucket does not exist.')
+        return f(bucket=bucket, *args, **kwargs)
+    return decorate
+
+
+def pass_object(f):
+    """Decorator to retrieve an object."""
+    @wraps(f)
+    def decorate(*args, **kwargs):
+        key = kwargs.pop('key')
+        obj = ObjectVersion.get(kwargs['bucket'].id, key)
+        if not obj:
+            obj = namedtuple('mock_obj', ('key'))(None)
+        return f(obj=obj, *args, **kwargs)
+    return decorate
+
+
+def pass_file(f):
+    """Decorator to retrieve uploaded file."""
+    @wraps(f)
+    def decorate(*args, **kwargs):
+        uploaded_file = request.files.get('file')
+        if not uploaded_file:
+            abort(400, 'File is missing from the request.')
+        return f(uploaded_file=uploaded_file, *args, **kwargs)
+    return decorate
+
+
+def need_bucket_collection_permission(action):
```

Comment:
Perhaps something along the lines of to avoid duplication of the needs_permission logic:

```python
def need_permissions(factory, action):
    """"Get permission for buckets or abort."""
    def decorator_builder(f):
        @wraps(f)
        def decorate(*args, **kwargs):
            permission = factory(kwargs, action)
            if permission is not None and not permission.can():
                if current_user.is_authenticated:
                    abort(403,
                          'You do not have a permission for this action')
                abort(401)
            return f(*args, **kwargs)
        return decorate
    return decorator_builder


need_bucket_collection_permission = partial(
    need_permission, 
    lambda kwargs, action: current_bucket_collection_permission_factory(action)
)

need_bucket_permission = partial(
    need_permission, 
    lambda kwargs, action: current_bucket_permission_factory(kwargs['bucket'], action=action)
)

need_objects_permission = partial(
    need_permission, 
    lambda kwargs, action: current_bucket_collection_permission_factory(kwargs['bucket'], kwargs['object'], action=action)
)
```